### PR TITLE
RC6 allow AoT Compilation using TemplateRefs

### PR DESCRIPTION
**Pull request by isaacplmann**
There are 2 big changes in this pull request.

1. There is now a TreeModule NgModule that should be familiar to everyone using RC5+.
2. treeNodeTemplate and loadingComponent options have been removed in favor of `<template>` tags in the content of the `Tree` component.

## Rational for change 2:
The main reason for the second change is that the ad-hoc component creation process wasn't going to work in AoT compilation.  This new method, while requiring a breaking change, will provide all the existing functionality and work AoT compilation.  (It also allows for some complex scenarios that aren't currently possible.)

## Breaking changes as a result of change 2:
Instead of passing template strings/components in the options object, templates will be created as `<template>` tags in the content of the `Tree` component.

### Before:
```html
<Tree [options]="options"></Tree>
```
```js
this.options = {
  loadingComponent: 'patience, young one...',
  treeNodeTemplate: '<a [href]="node.link">{{ node.text }}</a>',
}
```

### After:
```html
<Tree [options]="options">
  <template #treeNodeTemplate let-node>
    <a [href]="node.link">{{ node.text }}</a>
  </template>
  <template #loadingTemplate>
    patience, young one...
  </template>
</Tree>
```
```js
this.options = {
}
```

### After with components:
```html
<Tree [options]="options">
  <template #treeNodeTemplate let-node>
    Here's a component! <my-component [something]="node"></my-component>
  </template>
  <template #loadingTemplate>
    <patience-component></patience-component>, young one...
  </template>
</Tree>
```
```js
this.options = {
}
```

### Mixing contexts:

Coincidentally, this method also allows you to use values from the outer context within the tree node template.  Anything that isn't named `node` is available to be used as if it was outside the `Tree` component.

```html
<Tree [options]="options">
  <template #treeNodeTemplate let-node>
    inside: {{ node.text }} vs. outside: {{ outsideValue }}
  </template>
  <template #loadingTemplate>
    patience, young one...
  </template>
</Tree>
```
```js
this.outsideValue = "outside";
this.options = {
}
```

### CODE REVIEW

**Comment by adamkleingit** [comment]
I prefer to deprecate the old way (using the deprecated method in lib folder),
so people can still use the options method but they will get a warning in the console

**Comment by isaacplmann** [comment]
You're right. That's a better way to do it. I'll update the pull request
tomorrow.

**Comment by Gillardo** [comment]
All looks great to me...Does this update include the 100% width fixes as well??

**Comment by isaacplmann** [comment]
I've updated my pull request to continue support for the current configuration with deprecation warnings.  To allow for AoT compilation, I've isolated the deprecated functionality in DeprecatedTreeModule.  People who can't update their code should use that.

DeprecatedTreeModule will show a deprecation warning stating that AoT compilation will not be possible when using it.  It will also show deprecation warnings for options.treeNodeTemplate and options.loadingComponent, but the tree component will function the same as it did in the previous version.

TreeModule will show deprecation warnings for options.treeNodeTemplate and options.loadingComponent and ignore their values.

**Comment by adamkleingit** [comment]
OK Thanks.
I really hope to go over this in the coming days and merge.

-- 
- Adam Klein <http://il.linkedin.com/in/adamklein500>
- CTO @ 500Tech.com <http://500tech.com/>
- AngularJS, ReactJS, NodeJS
- Development, consulting, training and public speaking
- (+972) 52 747 5633

**Comment by sclausen** [comment]
I really like this :+1:

**Comment by isaacplmann** [comment]
Let me know if you have questions or concerns about the code.  I'm making it a priority to get this PR in a state that you're comfortable merging it.

**Comment by shlomiassaf** [comment]
@isaacplmann How did you get this to work?

You didn't change the source of the examples, did you smoke test it?

I tried running it  (dist) in my project and I get errors.

The compiled code doesn't seem to work with `@angular-cli/webpack`.

```
compiler.umd.js:14118 Uncaught Error: Unexpected value 'TreeModule' imported by the module 'AppModule'
``
 I belive due to https://github.com/angular/angular/issues/11438 

Anyway, IMO, this PR should include a followup upgrade to the webpack example otherwise the repo get's into an invalid state.

**Comment by shlomiassaf** [comment]
@adamkleingit I see a lot of potential for this library.

You should add a UMD bundle, I did it in my [modal](https://github.com/shlomiassaf/angular2-modal) library

**Comment by isaacplmann** [comment]
@shlomiassaf Yes, I ran it against my work app to make sure it worked.  Using webpack 2.1.0-beta.21 with a setup based off of AngularClass/angular2-webpack-starter.

You're right, I should update the examples.  I was holding off on that until I got some indication from @adamkleingit that he liked the changes.  It seems like he's leaning toward merging, so I'll spend some time updating the examples in the next few days.

File: lib/components/deprecated-tree-node-content.component.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+      componentFactory = this.componentFactoryResolver.resolveComponentFactory(this.treeModel.treeNodeContentComponent);
+    } catch(error) {
+      componentFactory = this.adHocComponentFactoryCreator.getFactory(this.treeModel.treeNodeContentComponent);
+    }
+    let componentRef: ComponentRef<ITreeNodeTemplate>
+      = this.viewContainerRef.createComponent<ITreeNodeTemplate>(componentFactory, 0, this.viewContainerRef.injector);
+    componentRef.instance.node = this.node;
+    componentRef.instance.context = this.node.context;
+
+    componentRef.changeDetectorRef.detectChanges();
```
Comment (adamkleingit) [inline lib/components/deprecated-tree-node-content.component.ts]: I didn't understand how does the changeDetectorRef get injected to the component?

File: lib/components/deprecated-loading.component.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  }
+
+  ngAfterViewInit() {
+    this._loadTreeNodeContent();
+  }
+
+  _loadTreeNodeContent() {
+    let componentFactory;
+    try {
+      componentFactory = this.componentFactoryResolver.resolveComponentFactory(this.treeModel.loadingComponent);
```
Comment (adamkleingit) [inline lib/components/deprecated-loading.component.ts]: Why do you try both method?

File: lib/models/tree-options.model.ts [hunk 3]
```diff
@@ -98,6 +96,14 @@ export class TreeOptions {
...
     }
 
+    if (options.treeNodeTemplate) {
+      deprecated('treeNodeTemplate', 'a template in the content of the <Tree> component like this: <Tree><template #treeNodeTemplate let-node>...</template></Tree>.  If you don\'t have time to update your code and don\'t need AoT compilation, use DeprecatedTreeModule');
+    }
+
+    if (options.loadingComponent) {
+      deprecated('loadingComponent', 'a template in the content of the <Tree> component like this: <Tree><template #loadingTemplate>...</template></Tree>.  If you don\'t have time to update your code and don\'t need AoT compilation, use DeprecatedTreeModule');
+    }
+
```
Comment (adamkleingit) [inline lib/models/tree-options.model.ts]: Now that we're working with templates we should also deprecate `options.context`.

**Comment by adamkleingit** [comment]
@isaacplmann - I went over PR. Made a few comments. I like the code a lot ;)

updating the examples will help a lot.
You can stick with updating the `webpack` example, I'll take care of
SystemJS.

To run the example, make sure you run `npm install` inside the webpack
folder, then
```
$ npm run example:webpack
```

Also - need to merge from master (I added filtering on the tree).

Thanks a lot for the contributions



-- 
- Adam Klein <http://il.linkedin.com/in/adamklein500>
- CTO @ 500Tech.com <http://500tech.com/>
- AngularJS, ReactJS, NodeJS
- Development, consulting, training and public speaking
- (+972) 52 747 5633

**Comment by shlomiassaf** [comment]
@adamkleingit rebase 😄

**Comment by adamkleingit** [comment]
Of course 😋

-- 
- Adam Klein <http://il.linkedin.com/in/adamklein500>
- CTO @ 500Tech.com <http://500tech.com/>
- AngularJS, ReactJS, NodeJS
- Development, consulting, training and public speaking
- (+972) 52 747 5633

File: lib/components/deprecated-tree-node-content.component.ts (hunk 1 above)
Comment (isaacplmann) [inline lib/components/deprecated-tree-node-content.component.ts]: changeDetectorRef comes as part of the componentRef object that is created by viewContainerRef.createComponent().  No need for us to do anything special to inject it.

File: lib/components/deprecated-loading.component.ts (hunk 2 above)
Comment (isaacplmann) [inline lib/components/deprecated-loading.component.ts]: I try componentFactoryResolver for the case where they have passed a component that is registered somewhere in their application.  If that fails, I know its an ad hoc component (created in tree-model.ts) that I need to instantiate with the adHocComponentFactoryCreator.

File: lib/models/tree-options.model.ts (hunk 3 above)
Comment (isaacplmann) [inline lib/models/tree-options.model.ts]: You're right.

**Comment by isaacplmann** [comment]
Webpack example done.  You'll need to change `example/webpack/package.json` to reference `../..` for `angular2-tree-component`.  So you'll do:

1. `npm run build` in the root folder
2. `npm install` in `example/webpack`
3. `npm run start` in `example/webpack`

Once these changes are published to npm, you won't need to modify the package.json to get the example to work.  And you'll be able to skip step 1 above.

**Comment by isaacplmann** [comment]
Ah.  I totally missed `npm run example:webpack`.  Looks like you already accounted for updating the node_modules in the example/webpack folder.  Ignore the `../..` comment above.

**Pull request merged**